### ISSUE TITLE
Entity merging works for non-existent keys but does nothing if key existed already.

### ISSUE THREAD

**Issue by petermetz** [Impetus/kundera#559]
Queries to read from the database are working and the merge as well, as long as I won't try to merge an entity with an existing key (which would be the point)
The entity I'm testing on is nothing special, just a bunch of elementary fields, no relationships defined, no embeddable, etc.
I tried to wrap the persist call between .begin() and .commit() transaction code as well.

If I change the entity's key to something random before calling merge (just for debugging purposes) then the new record gets created without any issues. If I don't change the entity key (therefore I'm trying to update) the changes won't be reflected in the database after a seemingly successful query. (neither Thrift neither Kundera reports any problems).

I tried to debug the issue and verified that Kundera calls the underlying batch_mutate() thrift method with the correct values (mutation map contains the correct values, timestamps are okay, etc). 
Based on this I think the configuration and the entities should be fine, moreover it might not even be a Kundera issue, but something with Thrift instead.

Any pointers on how to debug this would be helpful, and if it's a total low ball I'm happy to make the effort of putting together a full blown test case for it. (Just hoping that it is not necessary as it won't be a quick one)

Thanks in advance,
Peter

**Comment by mevivs** [comment]
Please create a sample junit for this. I suspect it is because of timestamp value.

-Vivek

**Comment by petermetz** [comment]
It was indeed the timestamp, sorry for giving a misleading explanation. I still have a problem though, namely that Kundera seems to hard code the mode of timestamp generation for mutations, and the legacy code (which I'm trying to replace with Kundera) does the same, but in a different manner.

Kundera uses `System.currentTimeMillis()` while the code that used to handle the data before uses
`System.currentTimeMillis() * 1000`.
And this causes any mutations (produced by Kundera) to seem outdated to the database.

Question: Is there any easy way to force/override Kundera's behaviour to comply with the mentioned timestamp generation behaviour?
If there is no such possibility right now, would you consider accepting a contribution which does so?
Not sure about how hard it would be but ideally it could be a property of the persistence unit which is definable in the persistence.xml file.

I'm closing the issue at the same time, since obviously there is nothing wrong with the Kundera code in this regard.

Thanks,
Peter

**Comment by mevivs** [comment]
```
Question: Is there any easy way to force/override Kundera's behaviour to comply with the mentioned timestamp generation behaviour?
```

Let me think about this and discuss with team. Will get back to you on this.

HTH,
-Vivek

**Comment by petermetz** [comment]
Thank you, very much appreciated!

Peter

**Comment by petermetz** [comment]
Hi Vivek,

Just wanted to ask if this still in discussion?

Thanks,
Peter

**Comment by mevivs** [comment]
Hi Peter,
Yes. We had a discussion, and need to finalize among given below points:
1. Provide a way to allow user plugin own TimeStampGenerator implementation.
2. Allow user to set time stamp per operation (similar to em.getDelegate() and then client.setConsistencyLevel).

With point 1, it would simply override default behavior, if implementation found in classpath. Question is what if multiple implementations found!

With point 2, certainly it would allow user to set timestamp per operation, but it's not one time setting.  Ideally we should allow both ways.

Will post an update once we finalize on implementation. Feel free to provide your inputs/suggestions.

Reopening, as with closed state somehow missed on this  :)

Thanks,
-Vivek

**Comment by petermetz** [comment]
Hi Vivek,

This was roughly my idea as well, just want to ask a quick question about point one:
Would the classpath scan take place based on annotations (let's say @TimeStampProvider) or it would check for a class which was declared in the persistence.xml? 
I guess the best would be to do both of them and let the user decide if they prefer annotations, or XML configuration, but that's more work as well.

If nothing found as a timestamp provider, the timestamps should be generated exactly the same way they are generated right now, to avoid confusion and breaking existing applications upon upgrading to a newer version.

Different consistency levels make sense as those can have a meaningful effect on the applications' performance, but I can't see why anybody would want to use a different timestamp generation method per request/column family or even datacenter. Even with all that said, it might be just my lack of imagination and if it is not hard to implement, why not allow the users to specify timestamp generation more flexibly. Personally, my use-case wouldn't need point 2.

Cheers,
Peter

**Comment by chhavigangwal** [comment]
Issue fixed. Released with Kundera-2.12.
